common/src/test/java/com/microsoft/identity/common/internal/ipc/mock/MockContextForBoundServiceConnectionFailed.java

```diff
@@ -0,0 +1,43 @@
+//  Copyright (c) Microsoft Corporation.
+//  All rights reserved.
+//
+//  This code is licensed under the MIT License.
+//
+//  Permission is hereby granted, free of charge, to any person obtaining a copy
+//  of this software and associated documentation files(the "Software"), to deal
+//  in the Software without restriction, including without limitation the rights
+//  to use, copy, modify, merge, publish, distribute, sublicense, and / or sell
+//  copies of the Software, and to permit persons to whom the Software is
+//  furnished to do so, subject to the following conditions :
+//
+//  The above copyright notice and this permission notice shall be included in
+//  all copies or substantial portions of the Software.
+//
+//  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+//  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+//  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+//  THE SOFTWARE.
+package com.microsoft.identity.common.internal.ipc.mock;
+
+import android.os.IInterface;
+
+import androidx.annotation.NonNull;
+
+import com.microsoft.identity.common.exception.ClientException;
+import com.microsoft.identity.common.internal.broker.BoundServiceClient;
+
+import org.robolectric.annotation.Implements;
+
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeoutException;
+
+@Implements(BoundServiceClient.class)
+public class MockContextForBoundServiceConnectionFailed<T extends IInterface> {
+    protected @NonNull T connect(@NonNull final String targetServicePackageName)
+            throws ClientException, InterruptedException, TimeoutException, ExecutionException {
+        throw new TimeoutException("Connection timed out");
```

Comment:
might be nice to parameterize this thing so you could make it throw them all by using a static factory that takes the exception to throw.

Reply:
can you parameterize a shadow class?

Reply:
I was thinking more about whether a shadow is the right way to deal with this, since BoundServiceClient is, I think, our own class.  We shouldn't *need* to use a shadow for it.

Reply:
we don't necessarily need shadow, but IMO it's easier/cleaner to implement this way.

Alternatively, we would have to supply IIpcStrategy object in every methods, across all 3 strategies. (right now each method are calling testOperationXXXXX(), and that function is pulling the strategy object from getStrategy()). Shadowing allows us to easily modify the injected object in specific test cases.